src/cpp/src/llm_pipeline.cpp

```diff
@@ -96,30 +94,30 @@ class StatefulLLMPipeline final : public LLMPipelineImplBase {
             encoded_input = m_tokenizer.encode(*input_vector);
         } else if (auto input_prompt = std::get_if<std::string>(&inputs)) {
             std::string& prompt = *input_prompt;
-            
+
             if (is_chat_conversation) {
-                m_history.push_back({{"role", "user"}, {"content", prompt}});
                 constexpr bool add_generation_prompt = true;
-                auto new_templated_chat_history  = m_tokenizer.apply_chat_template(m_history, add_generation_prompt);
-                
-                prompt = new_templated_chat_history.substr(m_templated_chat_history.size());
-                m_templated_chat_history = new_templated_chat_history;
+                ChatHistory history = {{{"role", "user"}, {"content", prompt}}};
+                prompt = m_tokenizer.apply_chat_template(history, add_generation_prompt);
+                // In case it's not the first prompt in chat conversation,
+                // need to append EOS token as this token already presented in KV-cache
+                if (!m_is_cache_empty) {
+                    prompt.insert(0, m_tokenizer.get_eos_token() + "\n");
```

Comment:
This might be the problem...

@Wovchena, @pavel-esir Can there be anything else except `<EOS>` token? (I doubt, but who knows)

Reply:
So. Given the last conversation, do you still want to handle `m_tokenizer.get_eos_token() + "\n"` only?

Reply:
No, the suffix can be different in general case. But still it's constant and can be calculated once.

Will update this part